### PR TITLE
Proposition to add support for empty responses

### PR DESCRIPTION
**Pull request by Pomanks**
This feature would let the client decide wether or not to handle some empty responses (e.g. some endpoint sometimes returning data, some other time nothing at all).

### CODE REVIEW

**Comment by philprime** [comment]
HI @Pomanks thanks for your PR. Before jumping into reviewing the changes in more detail, can you please provide an example on how to use your feature (example request definition + usage with client).

**Comment by Pomanks** [comment]
Hi @philprime, 

Sure thing! Here's a concrete example on how to use the feature with the given endpoint: https://trakt.docs.apiary.io/#reference/users/watching/get-watching

(As you can see in the endpoint's documentation, it sometimes returns some data and sometimes nothing at all.)

One way to handle this is to avoid the decoding to throw an error when no data is returned.

**Given the request:**

```swift 
struct GetWatching: Request {
    // Add your request parameters here…
}
```

**And its response:**

```swift
public extension GetWatching {

  struct Response: Decodable {

    @ResponseBody<Body> public var body
    @ResponseStatusCode public var statusCode
    // Since the data will be empty with a 204 status code, it can be useful to keep it for our internal logic to check wether our API call did return any data or not.
    // We can also just check if the response's body is nil since it now will when empty data is returned.

    struct Body: JSONDecodable {
      // ...
    }
  }
}
```

**We can use our request like so:**

```swift
func getWatching() -> AnyPublisher<GetWatching.Response, Error> {
  let request = GetWatching( ... )
    
  return httpClient.send(request, failsOnEmptyData: false) // Passing false prevents any decoding error to be thrown later on…
}
```

Please let me know if that's enough 🙂

**Comment by philprime** [comment]
Okay I see, let me rephrase and please correct me if I am wrong:

The endpoint has two possible response bodies:

1. Valid JSON Object in the scheme of `Body`, e.g.:

```json
{
    "some": "data"
}
```

2. Empty response body, therefore not even the smallest valid JSON `{}`.
This would lead to a decoding error, as `JSONDecoder` can't work with empty strings.

Is your PR a solution for the second one?

**Comment by Pomanks** [comment]
Precisely!

**Comment by philprime** [comment]
Great, I believe too that an optionally empty response body should be valid (especially with caching etc.)

Unfortunately there is a main issue with the implementation: Requests shall stay self-contained.

This is the main philosophy of Postie, the struct definition should contain all necessary info, an the client should not need additional parameters. Now if we remove the `failsOnEmptyData` parameter, we still somehow have to tell the `ResponseBody` decoding mechanism it should not fail with empty data.

We could add a flag to the `Request`, e.g. `@RequestShouldAllowEmptyResponses var shouldAllowEmptyResponse = true` but this seems verbose, and also the flag concerns the response, not the request.

Instead I have a different solution to offer: extending the `ResponseBody` property wrapper.

Unfortunately we can not provide additional information as values, as decoding doesn't allow it. Instead we can provide additional information to a decoding logic by using generics. First off we need to create a protocol with a static method which returns the flag we need:

```swift
public protocol ResponseBodyDecodingStrategy {

    static var shouldAcceptEmptyDataAsNil: Bool { get }

}
```

Now we add the decoding strategy to the `ResponseBody` struct:

```swift
@propertyWrapper
public struct ResponseBody<Body: Decodable, DecodingStrategy: ResponseBodyDecodingStrategy> {
    // ...
}
```

We can now access the static getter-method using the generic strategy:

```swift
extension ResponseBodyWrapper: Decodable {

    public init(from decoder: Decoder) throws {
        guard let responseDecoder = decoder as? ResponseDecoding else {
            self.wrappedValue = try Body(from: decoder)
            return
        }
        guard HTTPStatusCode.ok..<HTTPStatusCode.multipleChoices ~= responseDecoder.response.statusCode else {
            self.wrappedValue = nil
            return
        }
        if DecodingStrategy.shouldAcceptEmptyDataAsNil && responseDecoder.data.isEmpty {
            self.wrappedValue = nil
            return
        }
        self.wrappedValue = try responseDecoder.decodeBody(to: Body.self)
    }
}
```

We need a default strategy to keep the current logic:

```swift
extension DefaultStrategy: ResponseBodyDecodingStrategy {

    public static var shouldAcceptEmptyDataAsNil: Bool { return false }

}
```

and add a opt-in strategy to fit your proposed use-case:

```swift
public class AcceptEmptyAsNilStrategy: ResponseBodyDecodingStrategy {

    public static var shouldAcceptEmptyDataAsNil: Bool { true }
}
```

With this convenience type-aliases we can add more strategies as we go:

```swift
extension ResponseBody {

    public typealias AcceptEmptyAsNil = ResponseBodyWrapper<Body, AcceptEmptyAsNilStrategy>

}
```

For additional convenience, we change `ResponseBody<Body, DecodingStrategy>` to `ResponseBodyWrapper<Body, DecodingStrategy>` and add a new typealias `ResponseBody<Body>`:

```swift
public typealias ResponseBody<Body: Decodable> = ResponseBodyWrapper<Body, DefaultStrategy>
```

## Usage:

Original use-case requiring the response data:

```swift
struct Response: Decodable {

    struct Body: JSONDecodable {}

    @ResponseBody<Body> var body

}
```

Allowing empty as nil:

```swift
struct Response: Decodable {

    struct Body: JSONDecodable {}

    @ResponseBody<Body>.AcceptEmptyAsNil var body

}
```

and we can even use custom strategies, which we can extend further down the road, e.g. pass the response status code to the strategy to only accept empty responses for specific HTTP status codes.

```swift
public class MyCustomStrategy: ResponseBodyDecodingStrategy {

    public static var shouldAcceptEmptyDataAsNil: Bool {
        // e.g. only if the time is even (this is a stupid example)
        Date().timeIntervalSince1970.truncatingRemainder(dividingBy: 2) == 0
    }
}

struct Response: Decodable {

    struct Body: JSONDecodable {}

    @ResponseBodyWrapper<Body, MyCustomStrategy> var body

}
```

While tinkering I also pushed my changes on the branch `feature/response-body-stratgies` if you want to check it out.

What do you think about this approach?

**Comment by philprime** [comment]
Additionally, this solution would keep Postie clean as we would not have to pass the flag all the way down to the response body decoding

**Comment by Pomanks** [comment]
I think it's way better than my implementation and more Postie-like indeed! Mine was a lot “quick and dirty style” cause I needed to make it work first (and I'm not comfortable with property wrappers yet…) but yours is very clean.

I particularly love the custom strategy part “e.g. pass the response status code to the strategy to only accept empty responses for specific HTTP status codes.” since it directly solves one of my issues haha.

**Comment by philprime** [comment]
Great! I am not quite certain how the strategy methods should be defined, the `public static var shouldAcceptEmptyDataAsNil: Bool { return false }` was just an example.

How about you adapt your PR to also use the strategy pattern and adapt the `ResponseBodyDecodingStrategy` protocol so you can implement your "custom strategy part"? Maybe it is not as custom as you think and we can integrate part of it into the Postie core (e.g. 204 defaults to empty https://httpstatuses.com/204)

Edit:
if you want to know more about Annotating properties with default decoding values, John Sundell [has a great article](https://swiftbysundell.com/tips/default-decoding-values/)

**Comment by Pomanks** [comment]
I'll look into it, do you mind if I rebase my work from your branch `feature/response-body-strategies`?

Regarding status codes, it's even better if this 204 is the default for empty responses, I'll see what I can do for the custom strategy part.

And thanks for sharing the article 😄

**Comment by philprime** [comment]
Yes, you can go rebase on it (might need some cleanup tho).

Looking forward to an update :)

**Comment by Pomanks** [comment]
Updated! I also moved some stuff into folders (strategies mostly) to avoid having those declared all around.

Let me know what you think 🙂

**Comment by philprime** [comment]
Thanks for investing the time!

Looks good to me, a few things that came to my mind (mostly code-style):
- Please change the folder "Response Body" to something without spaces, I personally had issues in the past where escaping failed and it caused annoying problems.
- The `ResponseBody+EmptyBodyDecodingStrategy.swift` currently contains a single typealias. It probably doesn't make sense to create a new file for every new typealias, so let's rename it to `ResponseBody+DecodingStrategies.swift` or similar
- As we are creating multiple files anyways (i like it!) how about an own file for `ResponseBodyDecodingStrategy`?
- For convenience of the example code, I added the ResponseBodyDecodingStrategy to the `DefaultStrategy` struct, which was introduced for decoding headers. I am wondering if we should keep these default strategies separate, what do you think? Also because the `DefaultStrategyOptional` does not apply to the ResponseBody as it is Optional anyways (to support `@ResponseErrorBody`)
- `EmptyBodyAllowedStrategy` also implement `ResponseHeaderDecodingStrategy` now, which seems like a copy-paste error, as `EmptyBody` doesn't seem reasonable naming for response header strategies (might be confusing for future users).
- Also please fix the test case and add additional ones for FormURLEncoded body

**Comment by Pomanks** [comment]
> - Please change the folder "Response Body" to something without spaces, I personally had issues in the past where escaping failed and it caused annoying problems.

✅


> - The `ResponseBody+EmptyBodyDecodingStrategy.swift` currently contains a single typealias. It probably doesn't make sense to create a new file for every new typealias, so let's rename it to `ResponseBody+DecodingStrategies.swift` or similar 

✅

> - As we are creating multiple files anyways (i like it!) how about an own file for `ResponseBodyDecodingStrategy`?

✅ I also moved `ResponseHeaderDecodingStrategy` to its own separate file and inside Strategies folder.


> - For convenience of the example code, I added the `ResponseBodyDecodingStrategy` to the `DefaultStrategy` struct, which was introduced for decoding headers. I am wondering if we should keep these default strategies separate, what do you think? Also because the `DefaultStrategyOptional` does not apply to the `ResponseBody` as it is Optional anyways (to support `@ResponseErrorBody`)

Not sure where you added this (example code)?
Regarding the DefaultStrategies, I think we can kind of merge those two (or use something like `DefaultStrategy.Optional` but I'm not sure it fits the scope of the PR anymore haha

> - `EmptyBodyAllowedStrategy` also implement `ResponseHeaderDecodingStrategy` now, which seems like a copy-paste error, as EmptyBody doesn't seem reasonable naming for response header strategies (might be confusing for future users).

I wasn't sure if the `EmptyBodyAllowedStrategy` should behave the same as `DefaultStrategy` and it's why I copy-pasted this part…
My point being: even if the body is empty, we still might want to get the headers from the request's response, right?

Perhaps a better naming would be `OptionalContentStrategy` (it also makes more sense since the `body` will be `nil` when no content is found instead of empty like `"{}"`). What do you think? (I would also rename all matching protocols/typealiases accordingly)

> - Also please fix the test case and add additional ones for FormURLEncoded body

✅

**Comment by philprime** [comment]
> Not sure where you added this (example code)?

In my previous code sample I extended the `DefaultStrategy`

```swift
extension DefaultStrategy: ResponseBodyDecodingStrategy {

    public static var shouldAcceptEmptyDataAsNil: Bool { return false }

}
```

This was just for convience of this issue. 

> Regarding the DefaultStrategies, I think we can kind of merge those two (or use something like DefaultStrategy.Optional but I'm not sure it fits the scope of the PR anymore haha

Yes, I think the convenience types are out of scope right now.

What I mean now is, not sure if we should merge them, but instead use `DefaultHeaderStrategy` and `DefaultBodyStrategy` to keep them separate (I am thinking about the SoC separation-of-concern)

> I wasn't sure if the EmptyBodyAllowedStrategy should behave the same as DefaultStrategy and it's why I copy-pasted this part…

In general the `DefaultStrategy` is a WIP and will most likely be closely associated with the HTTP standards. Therefore, yes I think for 204 we could allow empty responses by default.

> My point being: even if the body is empty, we still might want to get the headers from the request's response, right?

That's exactly what I mean with separation of concern, those two shouldn't affect each other

> Perhaps a better naming would be OptionalContentStrategy (it also makes more sense since the body will be nil when no content is found instead of empty like "{}"). What do you think? (I would also rename all matching protocols/typealiases accordingly)

Oh I see your point, "empty (JSON) data" vs "empty HTTP body = 0 bytes". Interesting point, but yes let's go with the `OptionalContentStrategy`.

**Comment by Pomanks** [comment]
I separated both header and body strategies (as indicated in my commit message) in this PR since it was easier to integrate with the changes I made.

Hope it's what you expected 🤔

**Comment by philprime** [comment]
Yes, this is what I was expecting! Thank you :)

I just realized our `DecodingStrategy` currently has a flaw: the `statusCode` is a static variable, meaning we can't have multiple requests in parallel as they might interfere with each other. 
I believe we need to change it something like this:

```swift
public struct DefaultBodyStrategy: ResponseBodyDecodingStrategy {

    public static func allowsEmptyContent(statusCode: Int) -> Bool {
         return statusCode == 204
    }
}
```

**Comment by Pomanks** [comment]
You're absolutely right…!

Also, I've extended a bit the `RequestPathParameterValue` protocol to allow for `Int16`/`Int32`/`Int64` as well as default `Int` (to avoid having to cast them later on…)

**Comment by philprime** [comment]
I like this! Looks good from my side.
Do you think there is anything left do before merging?

**Comment by Pomanks** [comment]
I think for this PR's scope, we're good!
(I already extended the scope a bit much haha)

**Comment by philprime** [comment]
We'll yes the scope grew over time but this is perfectly fine for a 0.X.Y version hahaha
Thank you so much for your contribution. I'll merge this now

**Comment by Pomanks** [comment]
Glad I could help 😊
Thanks for helping me rework the very first draft and the PR along the way too!
